File: leetcode/python/0169_majority_element.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # return self.sorting(nums)
        return self.greedy(nums)
    
    def sorting(self, nums):
        nums.sort()
        if len(nums) == 1: return nums[0]

        currNum, countNum = nums[0], 0
        for i in range(0, len(nums)):
            if currNum != nums[i]:
                currNum = nums[i]
                countNum = 1
            else: countNum += 1

            if countNum > len(nums) // 2: return currNum

        return currNum
    
    def greedy(self, nums):
        count = 0
        for num in nums:
            if count == 0:
                candidate = num
                count += 1
            elif candidate == num: count += 1
            else: count -= 1
        return candidate

File: leetcode/java/0169_majority_element.java
class Solution {
    public int majorityElement(int[] nums) {
        if (nums.length <= 2) return nums[0];
        
        Arrays.sort(nums);
        int i = 0, num = nums[0], currOccur = 0;

        for (; i < nums.length; i++) {
            if (num != nums[i]) {
                num = nums[i];
                currOccur = 0;
            }
            currOccur++;

            if (currOccur > (nums.length / 2)) return nums[i];
        } 

        return nums[i - 1];
    }
}